File: algs4/dataStructure/LinkedStack.py
class Node(object):
    def __init__(self):
        self.item = None
        self.next = None

class Stack(object):
    '''
    这是Stack的链表实现
    '''

    def __init__(self):
        self.N = 0
        self.first = Node()
        self.compass = None

    def __iter__(self):
        return self

    def __next__(self):
        if self.compass == None:
            self.compass = self.first
        else:
            self.compass = self.compass.next

        if self.compass.next == None:
            raise StopIteration()

        return self.compass

    def isEmpty(self):
        return self.N == 0

    def size(self):
        return self.N

    def push(self, ele):
        oldfirst = self.first
        self.first = Node()
        self.first.item = ele
        self.first.next = oldfirst
        self.N += 1

    def pop(self):
        item = self.first.item
        self.first = self.first.next
        self.N -= 1
        return item

File: algs4.js/dataStructure/LinkedStack.js
/*
 * @Author: zhenbang.tang 
 * @Date: 2018-05-29 19:44:50 
 * @Last Modified by:   zhenbang.tang 
 * @Last Modified time: 2018-05-29 19:44:50 
 * 链表
 */

class Node {
  constructor() {
    this.item = null;
    this.next = null;
  }
}

class Stack {
  constructor() {
    this.N = 0;
    this.first = new Node();
  }

  [Symbol.iterator]() {
    let compass = null;
    return {
      next: () => {
        if (!compass) {
          compass = this.first;
        } else {
          compass = compass.next;
        }

        return {
          value: compass.item,
          done: !compass.next
        }
      }
    }
  }

  get isEmpty() {
    return this.N === 0;
  }

  get size() {
    return this.N;
  }

  push(ele) {
    const oldFirst = this.first;
    this.first = new Node();
    this.first.item = ele;
    this.first.next = oldFirst;
    this.N += 1;
  }

  pop() {
    const item = this.first.item;
    this.first = this.first.next;
    this.N -= 1;
    return item;
  }
}

const stack = new Stack();
stack.push(1);
stack.push(3);
stack.push(5);
console.log(...stack);